File: 963.最小面积矩形-ii.py
#
# @lc app=leetcode.cn id=963 lang=python3
#
# [963] 最小面积矩形 II
#

# @lc code=start
class Solution:
    def minAreaFreeRect(self, points: List[List[int]]) -> float:
        n = len(points)
        points_set = set(map(tuple, points))    # 加速查找 hash，空间换取时间
        # 所有的点都不同，每次取3个点，计算第4个点“应该”在哪里，“探”一下存不存在
        res = float('inf')
        for i in range(n):
            [x1, y1] = points[i]
            for j in range(n):
                if i == j:
                    continue
                [x2, y2] = points[j]
                for k in range(j + 1, n):
                    if k == i:
                        continue
                    [x3, y3] = points[k]
                    # p2 p3为副对角线，p1为一个顶角, 矩形的对角线对应点相加和相等
                    x4 = x2 + x3 - x1
                    y4 = y2 + y3 - y1
                    if (x4, y4) in points_set:          # “探”一下，第4个点，是否存在
                        v21 = (x2 - x1,  y2 - y1)
                        v31 = (x3 - x1,  y3 - y1)
                        if v21[0] * v31[0] + v21[1] * v31[1] == 0:  # 判断垂直。2个向量的内积为0
                            cur_area = (v21[0] ** 2 + v21[1] ** 2)**0.5 * \
                                (v31[0] ** 2 + v31[1] ** 2)**0.5
                            if cur_area < res:
                                res = cur_area
        return res if res != float('inf') else 0
# @lc code=end

File: 963.最小面积矩形-ii.go
/*
 * @lc app=leetcode.cn id=963 lang=golang
 *
 * [963] 最小面积矩形 II
 */

// @lc code=start
// func minAreaFreeRect(points [][]int) float64 {
// 	hash := map[[2]int]struct{}{}
// 	for _, p := range points {
// 		hash[[...]int{p[0], p[1]}] = struct{}{}
// 	}
// 	n := len(points)
// 	ans := 0.0
// 	for i := 0; i < n; i++ {
// 		for j := i + 1; j < n; j++ {
// 			for k := 0; k < n; k++ {
// 				if k == i || k == j {
// 					continue
// 				}
// 				if area, ok := getArea(points[i], points[j], points[k], hash); ok {
// 					if ans == 0.0 || ans > area {
// 						ans = area
// 					}
// 				}

// 			}
// 		}

// 	}
// 	return ans
// }

// func getArea(x, y, z []int, hash map[[2]int]struct{}) (float64, bool) {
// 	a := (x[1] - y[1]) * (y[1] - z[1])
// 	b := (x[0] - y[0]) * (y[0] - z[0])
// 	//a==-b 说明xy线段和yz 线段垂直,不垂直就不是矩形j
// 	if a != -b {
// 		return 0.0, false
// 	}
// 	//寻找例外一个点
// 	qx := x[0] + z[0] - y[0]
// 	qy := x[1] + z[1] - y[1]
// 	//如果不存在例外的点直接返回
// 	if _, ok := hash[[...]int{qx, qy}]; !ok {
// 		return 0.0, false
// 	}
// 	//斜边长度计算相乘就是面积
// 	area := math.Hypot(float64(x[1]-y[1]), float64(x[0]-y[0])) * math.Hypot(float64(y[1]-z[1]), float64(y[0]-z[0]))
// 	return area, true
// }
func minAreaFreeRect(points [][]int) float64 {
	hash := map[[2]int]struct{}{}
	for _, point := range points {
		hash[[2]int{point[0], point[1]}] = struct{}{}
	}
	n := len(points)
	res := math.MaxFloat64
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			for k := 0; k < n; k++ {
				if k == i || k == j { // 点不能重合
					continue
				}
				// 根据矩形对角点之和相等计算出对应的第四个点坐标
				currNeed := [2]int{points[j][0] + points[k][0] - points[i][0], points[j][1] + points[k][1] - points[i][1]}
				if _, ok := hash[currNeed]; ok { // 如果第四个点在字典中存在
					// 对应的点存在
					vx0 := points[j][0] - points[i][0]
					vx1 := points[j][1] - points[i][1]
					vy0 := points[k][0] - points[i][0]
					vy1 := points[k][1] - points[i][1]

					if vx0*vy0+vx1*vy1 == 0 { // 根据向量积判定是否是一个真正的矩形
						// 向量积为0
						// 计算出一个点的两条邻边，矩形的话乘起来就好
						currArea := math.Sqrt(float64(vx0*vx0+vx1*vx1)) * math.Sqrt(float64(vy0*vy0+vy1*vy1))
						if currArea < res {
							// fmt.Println(points[i],points[j],points[k])
							// fmt.Println("width:", float64(vx0*vx0 + vx1 * vx1))
							// fmt.Println("height:", float64(vy0 * vy0 + vy1*vy1))
							res = currArea
						}
					}
				}
			}
		}
	}
	if res == math.MaxFloat64 { // 未出现最小值则返回 0
		return 0.0
	} else {
		return res // 出现了则返回最小值
	}
}

// @lc code=end

